clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java

```diff
@@ -370,6 +378,200 @@ private static DefaultRecord readFrom(ByteBuffer buffer,
+                        int stepsToLeftShift = buffer.position();
+                        int bytesToLeftShift = buffer.remaining();
+                        for (int i = 0; i < bytesToLeftShift; i++) {
+                            array[i] = array[i + stepsToLeftShift];
+                        }
+
+                        // then try to read more bytes to the remaining of the array
+                        int bytesToRead = Math.min(bytesRemaining, size - bytesToLeftShift);
+                        input.readFully(array, bytesToLeftShift, bytesToRead);
+                        bytesRemaining = bytesRemaining - bytesToRead;
+
+                        buffer = ByteBuffer.wrap(array);
+                        // only those many bytes are readable
+                        buffer.limit(bytesToLeftShift + bytesToRead);
+
+                        needMore = false;
+                    } else {
+                        throw new InvalidRecordException("Invalid record size: expected to read more bytes in record payload");
+                    }
+                }
+
+                switch (state) {
+                    case READ_KEY_SIZE:
+                        // read key size (varint, max 4bytes)
+                        if (buffer.remaining() < 4 && bytesRemaining > 0) {
+                            needMore = true;
+                        } else {
+                            keySize = ByteUtils.readVarint(buffer);
+                            bytesRead += ByteUtils.sizeOfVarint(keySize);
+                            if (keySize > 0) {
+                                if (remainingHeadersToSkip < 0) {
+                                    recordKeySize = keySize;
+                                }
+                                state = ReadState.SKIP_KEY;
+                                remainingKeyBytesToSkip = keySize;
+                            } else {
+                                state = ReadState.READ_VALUE_SIZE;
+                            }
+                        }
+                        break;
+
+                    case SKIP_KEY:
+                        // skip until key is consumed
+                        if (remainingKeyBytesToSkip > buffer.remaining()) {
+                            remainingKeyBytesToSkip -= buffer.remaining();
+                            buffer.position(buffer.limit());
+                            needMore = true;
+                        } else {
+                            buffer.position(buffer.position() + remainingKeyBytesToSkip);
+                            bytesRead += keySize;
+                            state = ReadState.READ_VALUE_SIZE;
+                        }
+                        break;
+
+                    case READ_VALUE_SIZE:
+                        // read value size (varint, max 4bytes)
+                        if (buffer.remaining() < 4 && bytesRemaining > 0) {
```

Comment:
I think varints are max 5 bytes, right?

Reply:
Ah good point! Also verified it is 5bytes